### ISSUE TITLE
fused LSTM op doesn't support double backward

### ISSUE THREAD

**Issue by ssnl** [pytorch/pytorch#9229]
and users can't get around it :/ 

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @mruberry @jbschlosser

**Comment by zou3519** [comment]
I have some skeleton code for what It should look like... I can try to flush that out

**Comment by albanD** [comment]
@zou3519 do I remember correctly here that the main issue is that implementation like cudnn do not provide double backward and don't provide the right intermediates for us to compute that double backward either? And so we would need to do something like this to support it: https://github.com/pytorch/pytorch/blob/aa18d174553e1c2ab4d9e09ae45c6ac323f04af5/tools/autograd/derivatives.yaml#L798

**Comment by zou3519** [comment]
Yes, the problem is that cudnn doesn't provide double backwards and doesn't provide the correct intermediates for us to compute the double backwards. In order to compute double backwards for RNNs, one has to make sure they are not using the cudnn implementation (the fused LSTMCell RNN implementation does support double backward).

The easiest thing to do here is document how to compute double backwards when using RNNs by turning off cudnn (and suggesting that there will be a performance hit, especially if the RNN has multiple layers).

Alternatively, we would need some sort of API where the user tells us "hey, I would like to compute the double backward of this RNN, thank you" when the user runs the RNN's forward pass. That could look like the following:
- a flag that gets passed to the RNN's forward pass
- a flag that gets passed to the RNN's constructor
- a context manager(?)

**Comment by albanD** [comment]
What the example I gave above is doing is checking, when the backward is running, if grad mode is enabled, the run a backward that is differentiable. Otherwise, run the non-differentiable one. I think that would work here as well no?

**Comment by zou3519** [comment]
> What the example I gave above is doing is checking, when the backward is running, if grad mode is enabled, the run a backward that is differentiable. Otherwise, run the non-differentiable one. I think that would work here as well no?

We are unable to run an efficient differentiable backward for LSTM if the forward computation went through cudnn. This because during the forward pass, cudnn saves intermediate values into some buffers and we do not know how those buffers work (it is an implementation detail for cudnn that is not supposed to be a stable API). 

We could just re-run the forward pass without cudnn during the backward pass to get some intermediate values though.

**Comment by albanD** [comment]
Ho ok, so that is not an option.
I guess that raising a nice error that explicitly says that you should run the forward without cudnn if you need double backward is a good solution?

Also we already have such a context manager:
```python
import torch

print(torch.backends.cudnn.enabled)
with torch.backends.cudnn.flags(enabled=False):
    print(torch.backends.cudnn.enabled)
print(torch.backends.cudnn.enabled)
```
will print
```
True
False
True
```

**Comment by zou3519** [comment]
> I guess that raising a nice error that explicitly says that you should run the forward without cudnn if you need double backward is a good solution?

Agree! I tried to do this in https://github.com/pytorch/pytorch/pull/33884 but it caused tests to fail wildly. I should go back and investigate.

> Also we already have such a context manager:

Nice! It will be helpful to add that to the error message PR (it looks like I did mention it in the error message but forgot about it until now haha)

Speaking of which, this issue is probably a duplicate of https://github.com/pytorch/pytorch/issues/5261 (depends on what we are talking about when we say "fused LSTM op").
